File: java/src/test/java/com/github/copilot/E2ETestContext.java
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------------------------------------------*/

package com.github.copilot;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.logging.Logger;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import com.github.copilot.rpc.CopilotClientOptions;

/**
 * E2E test context that manages the test environment including the CapiProxy,
 * working directories, and CLI path.
 *
 * <p>
 * This provides a complete test environment similar to the Node.js, .NET, Go,
 * and Python SDK test harnesses. It manages:
 * </p>
 * <ul>
 * <li>A replaying CapiProxy for deterministic API responses</li>
 * <li>Temporary home and work directories for test isolation</li>
 * <li>Environment variables for the Copilot CLI</li>
 * </ul>
 *
 * <p>
 * Usage example:
 * </p>
 *
 * <pre>
 * {@code
 * try (E2ETestContext ctx = E2ETestContext.create()) {
 * 	ctx.configureForTest("tools", "my_test_name");
 *
 * 	try (CopilotClient client = ctx.createClient()) {
 * 		CopilotSession session = client
 * 				.createSession(new SessionConfig().setOnPermissionRequest(PermissionHandler.APPROVE_ALL)).get();
 * 		// ... run test ...
 * 	}
 * }
 * }
 * </pre>
 */
public class E2ETestContext implements AutoCloseable {

    private static final Logger LOG = Logger.getLogger(E2ETestContext.class.getName());

    /**
     * The default GitHub token used by the CLI in e2e tests. The proxy resolves
     * this token to the default Copilot user registered at context creation.
     */
    private static final String DEFAULT_GITHUB_TOKEN = "fake-token-for-e2e-tests";
    private static final Pattern SNAKE_CASE = Pattern.compile("[^a-zA-Z0-9]");
    private static final Pattern USER_CONTENT_PATTERN = Pattern
            .compile("^\\s+-\\s+role:\\s+user\\s*$\\s+content:\\s*(.+?)$", Pattern.MULTILINE);

    private final String cliPath;
    private final Path homeDir;
    private final Path workDir;
    private String proxyUrl;
    private final CapiProxy proxy;
    private final Path repoRoot;
    private Path currentSnapshotFile;

    private E2ETestContext(String cliPath, Path homeDir, Path workDir, String proxyUrl, CapiProxy proxy,
            Path repoRoot) {
        this.cliPath = cliPath;
        this.homeDir = homeDir;
        this.workDir = workDir;
        this.proxyUrl = proxyUrl;
        this.proxy = proxy;
        this.repoRoot = repoRoot;
    }

    /**
     * Creates a new E2E test context.
     *
     * @return the test context
     * @throws IOException
     *             if setup fails
     * @throws InterruptedException
     *             if setup is interrupted
     */
    public static E2ETestContext create() throws IOException, InterruptedException {
        Path repoRoot = findRepoRoot();
        String cliPath = getCliPath(repoRoot);

        Path tempDir = Paths.get(System.getProperty("java.io.tmpdir"));
        Path homeDir = Files.createTempDirectory(tempDir, "copilot-test-config-");
        Path workDir = Files.createTempDirectory(tempDir, "copilot-test-work-");

        CapiProxy proxy = new CapiProxy();
        String proxyUrl = proxy.start();

        // Register a default Copilot user for the CLI's default token so the proxy's
        // /copilot_internal/user endpoint returns a valid user (HTTP 200) instead of
        // 401 "Bad credentials". CLI 1.0.64-1 gates MCP enablement on this user:
        // `is_mcp_enabled` (added by the proxy) is the global gate, and snake_case
        // `copilot_plan` makes the third-party MCP policy resolver early-return
        // allow-all for non-org plans (anything other than business/enterprise),
        // avoiding a /copilot/mcp_registry network call the proxy does not serve.
        // Without this, MCP servers never reach CONNECTED. This mirrors the Go,
        // Node, Python, and .NET harnesses, which all register the same default
        // individual_pro user at context creation.
        Map<String, Object> defaultUser = new HashMap<>();
        defaultUser.put("login", "e2e-test-user");
        defaultUser.put("copilot_plan", "individual_pro");
        defaultUser.put("endpoints", Map.of("api", proxyUrl, "telemetry", "https://localhost:1/telemetry"));
        defaultUser.put("analytics_tracking_id", "e2e-test-tracking-id");
        proxy.setCopilotUserByToken(DEFAULT_GITHUB_TOKEN, defaultUser);

        return new E2ETestContext(cliPath, homeDir, workDir, proxyUrl, proxy, repoRoot);
    }

    /**
     * Gets the Copilot CLI path.
     */
    public String getCliPath() {
        return cliPath;
    }

    /**
     * Gets the temporary home directory for test isolation.
     */
    public Path getHomeDir() {
        return homeDir;
    }

    /**
     * Gets the temporary working directory for tests.
     */
    public Path getWorkDir() {
        return workDir;
    }

    /**
     * Gets the repository root for locating shared test assets.
     */
    public Path getRepoRoot() {
        return repoRoot;
    }

    /**
     * Gets the proxy URL.
     */
    public String getProxyUrl() {
        return proxyUrl;
    }

    /**
     * Configures the proxy for a specific test.
     *
     * @param testFile
     *            the test category folder (e.g., "tools", "session", "permissions")
     * @param testName
     *            the test method name (will be converted to snake_case)
     * @throws IOException
     *             if configuration fails
     * @throws InterruptedException
     *             if configuration is interrupted
     */
    public void configureForTest(String testFile, String testName) throws IOException, InterruptedException {
        // Restart the proxy if it has crashed
        ensureProxyAlive();

        // Convert test method names to lowercase snake_case for snapshot filenames
        // to avoid case collisions on case-insensitive filesystems (macOS/Windows)
        String sanitizedName = SNAKE_CASE.matcher(testName).replaceAll("_").toLowerCase();
        Path snapshotFile = repoRoot.resolve("test").resolve("snapshots").resolve(testFile)
                .resolve(sanitizedName + ".yaml");

        // Validate snapshot exists - fail fast with a clear message
        if (!Files.exists(snapshotFile)) {
            Path snapshotsDir = repoRoot.resolve("test").resolve("snapshots").resolve(testFile);
            String availableSnapshots = "";
            if (Files.exists(snapshotsDir)) {
                try (var files = Files.list(snapshotsDir)) {
                    availableSnapshots = files.filter(p -> p.toString().endsWith(".yaml"))
                            .map(p -> p.getFileName().toString().replace(".yaml", "")).sorted()
                            .reduce((a, b) -> a + ", " + b).orElse("<none>");
                }
            }
            throw new IOException(String.format(
                    "Snapshot file not found: %s%n" + "Category: %s, Test: %s (sanitized: %s)%n"
                            + "Available snapshots in '%s/': %s%n"
                            + "Ensure the snapshot exists and the test name matches exactly.",
                    snapshotFile, testFile, testName, sanitizedName, testFile, availableSnapshots));
        }

        this.currentSnapshotFile = snapshotFile;
        proxy.configure(snapshotFile.toString(), workDir.toString());

        // Log expected prompts to help debug prompt mismatch issues
        List<String> expectedPrompts = getExpectedUserPrompts();
        if (!expectedPrompts.isEmpty()) {
            LOG.info(() -> String.format("Configured snapshot '%s/%s' expects prompts: %s", testFile, sanitizedName,
                    expectedPrompts));
        }
    }

    /**
     * Gets the expected user prompts from the current snapshot file.
     * <p>
     * This is useful for debugging when tests fail with "No cached response found"
     * errors from CapiProxy. The prompts in your test must match these exactly.
     * </p>
     *
     * @return list of expected user prompt strings, or empty list if none found
     */
    public List<String> getExpectedUserPrompts() {
        if (currentSnapshotFile == null || !Files.exists(currentSnapshotFile)) {
            return List.of();
        }
        try {
            String content = Files.readString(currentSnapshotFile);
            List<String> prompts = new ArrayList<>();
            Matcher matcher = USER_CONTENT_PATTERN.matcher(content);
            while (matcher.find()) {
                String prompt = matcher.group(1).trim();
                // Remove quotes if present
                if ((prompt.startsWith("\"") && prompt.endsWith("\""))
                        || (prompt.startsWith("'") && prompt.endsWith("'"))) {
                    prompt = prompt.substring(1, prompt.length() - 1);
                }
                if (!prompts.contains(prompt)) {
                    prompts.add(prompt);
                }
            }
            return prompts;
        } catch (IOException e) {
            LOG.warning("Failed to read snapshot file: " + e.getMessage());
            return List.of();
        }
    }

    /**
     * Ensures the proxy is alive, restarting it if necessary.
     *
     * @throws IOException
     *             if the proxy cannot be restarted
     * @throws InterruptedException
     *             if interrupted during restart
     */
    public void ensureProxyAlive() throws IOException, InterruptedException {
        if (!proxy.isAlive()) {
            proxyUrl = proxy.restart();
        }
    }

    /**
     * Gets the captured HTTP exchanges from the proxy.
     *
     * @return list of exchange maps
     * @throws IOException
     *             if the request fails
     * @throws InterruptedException
     *             if the request is interrupted
     */
    public List<Map<String, Object>> getExchanges() throws IOException, InterruptedException {
        return proxy.getExchanges();
    }

    /**
     * Gets the environment variables needed for the Copilot CLI.
     *
     * @return map of environment variables
     */
    public Map<String, String> getEnvironment() {
        Map<String, String> env = new HashMap<>(System.getenv());
        env.put("COPILOT_API_URL", proxyUrl);
        // Route GitHub API calls (e.g. the MCP registry policy check) to the
        // replay proxy so MCP enablement stays hermetic. Without this the CLI
        // reaches the real api.github.com, which is slow/unreachable on macOS
        // CI runners and makes MCP servers time out before reaching connected.
        env.put("COPILOT_DEBUG_GITHUB_API_URL", proxyUrl);
        env.put("COPILOT_HOME", homeDir.toString());
        env.put("GH_CONFIG_DIR", homeDir.toString());
        env.put("XDG_CONFIG_HOME", homeDir.toString());
        env.put("XDG_STATE_HOME", homeDir.toString());

        // Configure CONNECT proxy for HTTPS interception if available
        String connectUrl = proxy.getConnectProxyUrl();
        String caFile = proxy.getCaFilePath();
        if (connectUrl != null && !connectUrl.isEmpty() && caFile != null && !caFile.isEmpty()) {
            String noProxy = "127.0.0.1,localhost,::1";
            env.put("HTTP_PROXY", connectUrl);
            env.put("HTTPS_PROXY", connectUrl);
            env.put("http_proxy", connectUrl);
            env.put("https_proxy", connectUrl);
            env.put("NO_PROXY", noProxy);
            env.put("no_proxy", noProxy);
            env.put("NODE_EXTRA_CA_CERTS", caFile);
            env.put("SSL_CERT_FILE", caFile);
            env.put("REQUESTS_CA_BUNDLE", caFile);
            env.put("CURL_CA_BUNDLE", caFile);
            env.put("GIT_SSL_CAINFO", caFile);
            env.put("GH_TOKEN", DEFAULT_GITHUB_TOKEN);
            env.put("GITHUB_TOKEN", DEFAULT_GITHUB_TOKEN);
            env.put("GH_ENTERPRISE_TOKEN", "");
            env.put("GITHUB_ENTERPRISE_TOKEN", "");
        }

        return env;
    }

    /**
     * Creates a CopilotClient configured for this test context.
     *
     * @return a new CopilotClient
     */
    public CopilotClient createClient() {
        CopilotClientOptions options = new CopilotClientOptions().setCliPath(cliPath).setCwd(workDir.toString())
                .setEnvironment(getEnvironment()).setGitHubToken(DEFAULT_GITHUB_TOKEN);

        return new CopilotClient(options);
    }

    /**
     * Creates a CopilotClient with the given options, applied on top of the default
     * options for this test context.
     *
     * @param options
     *            options to apply; environment and cliPath will be set from the
     *            context if not already set
     * @return a new CopilotClient
     */
    public CopilotClient createClient(CopilotClientOptions options) {
        if (options.getCliPath() == null) {
            options.setCliPath(cliPath);
        }
        if (options.getCwd() == null) {
            options.setCwd(workDir.toString());
        }
        if (options.getEnvironment() == null || options.getEnvironment().isEmpty()) {
            options.setEnvironment(getEnvironment());
        }
        if (options.getGitHubToken() == null) {
            options.setGitHubToken(DEFAULT_GITHUB_TOKEN);
        }

        return new CopilotClient(options);
    }

    /**
     * Configures the proxy to return a specific Copilot user response for a given
     * token. Used for per-session authentication tests.
     *
     * @param token
     *            the GitHub token
     * @param login
     *            the user login
     * @param copilotPlan
     *            the Copilot plan
     * @param apiUrl
     *            the API URL for the user endpoints
     * @param telemetryUrl
     *            the telemetry URL
     * @param analyticsTrackingId
     *            the analytics tracking ID
     * @throws IOException
     *             if the request fails
     * @throws InterruptedException
     *             if the request is interrupted
     */
    public void setCopilotUserByToken(String token, String login, String copilotPlan, String apiUrl,
            String telemetryUrl, String analyticsTrackingId) throws IOException, InterruptedException {
        ensureProxyAlive();
        proxy.setCopilotUserByToken(token, login, copilotPlan, apiUrl, telemetryUrl, analyticsTrackingId);
    }

    /**
     * Initializes the proxy state without loading a snapshot.
     * <p>
     * Use this for tests that need the proxy to be active (e.g., for per-session
     * auth token resolution via {@code /copilot_internal/user}) but do not make AI
     * completion requests and therefore have no snapshot to load.
     * </p>
     * <p>
     * The proxy requires its internal {@code state} to be initialized before it can
     * handle most endpoints. Without this call the proxy throws an error and
     * returns HTTP 500 for any request that arrives before a {@code /config} POST
     * has been made.
     * </p>
     *
     * @throws IOException
     *             if the proxy configuration request fails
     * @throws InterruptedException
     *             if the request is interrupted
     */
    public void initializeProxy() throws IOException, InterruptedException {
        ensureProxyAlive();
        // Pass a non-existent snapshot path. The proxy initializes its state even when
        // the file is absent (storedData simply remains undefined), which is fine for
        // tests that never make AI chat-completion requests.
        proxy.configure(workDir.resolve("no-snapshot.yaml").toString(), workDir.toString());
    }

    @Override
    public void close() throws Exception {
        proxy.stop();

        // Clean up temp directories (best effort)
        deleteRecursively(homeDir);
        deleteRecursively(workDir);
    }

    private static Path findRepoRoot() throws IOException {
        // First, check for copilot.sdk.dir system property (set by Maven during tests)
        String sdkDir = System.getProperty("copilot.sdk.dir");
        if (sdkDir != null && !sdkDir.isEmpty()) {
            Path sdkPath = Paths.get(sdkDir);
            if (Files.exists(sdkPath)) {
                return sdkPath;
            }
        }

        // Fallback: search up from current directory
        Path dir = Paths.get(System.getProperty("user.dir"));
        while (dir != null) {
            if (Files.exists(dir.resolve("nodejs")) && Files.exists(dir.resolve("test").resolve("harness"))) {
                return dir;
            }
            dir = dir.getParent();
        }
        throw new IOException("Could not find repository root. Either set copilot.sdk.dir system property "
                + "or run from within the copilot-sdk repository.");
    }

    private static String getCliPath(Path repoRoot) throws IOException {
        // Try environment variable first (explicit override)
        String envPath = System.getenv("COPILOT_CLI_PATH");
        if (envPath != null && !envPath.isEmpty()) {
            return envPath;
        }

        // Try test harness platform-specific binary (preferred as it has correct
        // version)
        String os = System.getProperty("os.name").toLowerCase();
        String arch = System.getProperty("os.arch").toLowerCase();
        String platform = os.contains("mac") ? "darwin" : os.contains("win") ? "win32" : "linux";
        String cpuArch = arch.contains("aarch64") || arch.contains("arm64") ? "arm64" : "x64";
        Path platformBinary = repoRoot
                .resolve("test/harness/node_modules/@github/copilot-" + platform + "-" + cpuArch + "/copilot");
        if (os.contains("win")) {
            platformBinary = repoRoot
                    .resolve("test/harness/node_modules/@github/copilot-" + platform + "-" + cpuArch + "/copilot.exe");
        }
        if (Files.exists(platformBinary)) {
            return platformBinary.toString();
        }

        // Try test harness npm-loader.js
        Path harnessCliPath = repoRoot.resolve("test/harness/node_modules/@github/copilot/npm-loader.js");
        if (Files.exists(harnessCliPath)) {
            return harnessCliPath.toString();
        }

        // Try nodejs installation. As of CLI 1.0.64-1 the @github/copilot package
        // is a thin loader; the runnable index.js ships in the installed
        // platform-specific package (e.g. @github/copilot-linux-x64). Exactly one
        // is installed. Running index.js under Node.js is the documented preferred
        // entry point and matches the Go, Python, Rust, and .NET test harnesses.
        Path githubModules = repoRoot.resolve("nodejs/node_modules/@github");
        if (Files.isDirectory(githubModules)) {
            try (var modules = Files.newDirectoryStream(githubModules, "copilot-*")) {
                for (Path module : modules) {
                    Path indexJs = module.resolve("index.js");
                    if (Files.exists(indexJs)) {
                        return indexJs.toString();
                    }
                }
            }
        }

        // Fallback: try to find 'copilot' in PATH
        String copilotInPath = findCopilotInPath();
        if (copilotInPath != null) {
            return copilotInPath;
        }

        throw new IOException("CLI not found. Either install 'copilot' globally, set COPILOT_CLI_PATH, "
                + "or run 'npm install' in the nodejs directory or test/harness directory.");
    }

    private static String findCopilotInPath() {
        try {
            String command = System.getProperty("os.name").toLowerCase().contains("win") ? "where" : "which";
            ProcessBuilder pb = new ProcessBuilder(command, "copilot");
            pb.redirectErrorStream(true);
            Process process = pb.start();
            try (BufferedReader reader = new BufferedReader(new InputStreamReader(process.getInputStream()))) {
                String line = reader.readLine();
                int exitCode = process.waitFor();
                if (exitCode == 0 && line != null && !line.isEmpty()) {
                    return line.trim();
                }
            }
        } catch (Exception e) {
            // Ignore - copilot not found in PATH
        }
        return null;
    }

    private static void deleteRecursively(Path path) {
        try {
            if (Files.exists(path)) {
                Files.walk(path).sorted((a, b) -> b.compareTo(a)) // Reverse order to delete children first
                        .forEach(p -> {
                            try {
                                Files.delete(p);
                            } catch (IOException e) {
                                // Best effort
                            }
                        });
            }
        } catch (IOException e) {
            // Best effort
        }
    }
}

File: dotnet/test/Harness/E2ETestContext.cs
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------------------------------------------*/

using Microsoft.Extensions.Logging;
using System.Diagnostics;
using System.Runtime.CompilerServices;
using System.Text.RegularExpressions;

namespace GitHub.Copilot.Test.Harness;

public sealed class E2ETestContext : IAsyncDisposable
{
    private const string DefaultGitHubToken = "fake-token-for-e2e-tests";

    public string HomeDir { get; }
    public string WorkDir { get; }
    public string ProxyUrl { get; }

    /// <summary>Optional logger injected by tests; applied to all clients created via <see cref="CreateClient"/>.</summary>
    public ILogger? Logger { get; set; }

    private readonly CapiProxy _proxy;
    private readonly string _repoRoot;
    private readonly object _clientsLock = new();
    private readonly List<CopilotClient> _persistentClients = [];
    private readonly List<CopilotClient> _transientClients = [];

    private E2ETestContext(string homeDir, string workDir, string proxyUrl, CapiProxy proxy, string repoRoot)
    {
        HomeDir = homeDir;
        WorkDir = workDir;
        ProxyUrl = proxyUrl;
        _proxy = proxy;
        _repoRoot = repoRoot;
    }

    public static async Task<E2ETestContext> CreateAsync()
    {
        var repoRoot = FindRepoRoot();

        var homeDir = Path.Combine(Path.GetTempPath(), $"copilot-test-config-{Guid.NewGuid()}");
        var workDir = Path.Combine(Path.GetTempPath(), $"copilot-test-work-{Guid.NewGuid()}");

        Directory.CreateDirectory(homeDir);
        Directory.CreateDirectory(workDir);

        // Resolve symlinks (e.g., macOS /var -> /private/var) so paths
        // match what spawned subprocesses see when they resolve their cwd.
        homeDir = ResolveSymlinks(homeDir);
        workDir = ResolveSymlinks(workDir);

        var proxy = new CapiProxy();
        var proxyUrl = await proxy.StartAsync();
        await proxy.SetCopilotUserByTokenAsync(DefaultGitHubToken, new CopilotUserConfig(
            Login: "e2e-test-user",
            CopilotPlan: "individual_pro",
            Endpoints: new CopilotUserEndpoints(Api: proxyUrl, Telemetry: "https://localhost:1/telemetry"),
            AnalyticsTrackingId: "e2e-test-tracking-id"));

        return new E2ETestContext(homeDir, workDir, proxyUrl, proxy, repoRoot);
    }

    /// <summary>
    /// Returns a canonical path with symlinks resolved in every directory
    /// component. .NET has no built-in equivalent of POSIX <c>realpath</c>
    /// that walks all parents, so we walk the components ourselves and use
    /// <see cref="DirectoryInfo.ResolveLinkTarget(bool)"/> on each one.
    /// On Windows, where the test temp paths don't traverse symlinks,
    /// <see cref="Path.GetFullPath(string)"/> is sufficient.
    /// </summary>
    private static string ResolveSymlinks(string path)
    {
        if (OperatingSystem.IsWindows())
        {
            return Path.GetFullPath(path);
        }

        try
        {
            var fullPath = Path.GetFullPath(path);
            var root = Path.GetPathRoot(fullPath);
            if (string.IsNullOrEmpty(root))
            {
                return fullPath;
            }

            var components = fullPath
                .Substring(root.Length)
                .Split(Path.DirectorySeparatorChar, StringSplitOptions.RemoveEmptyEntries);

            var resolved = root;
            foreach (var component in components)
            {
                resolved = Path.Join(resolved, component);
                try
                {
                    var info = new DirectoryInfo(resolved);
                    if (info.Exists && info.LinkTarget != null)
                    {
                        var target = info.ResolveLinkTarget(returnFinalTarget: true);
                        if (target != null && !string.IsNullOrEmpty(target.FullName))
                        {
                            resolved = target.FullName;
                        }
                    }
                }
                catch (Exception ex) when (ex is IOException or UnauthorizedAccessException)
                {
                    // Component we can't inspect; keep what we have and continue.
                }
            }

            return resolved;
        }
        catch (Exception ex) when (ex is IOException
            or UnauthorizedAccessException
            or ArgumentException
            or NotSupportedException
            or PathTooLongException)
        {
            return Path.GetFullPath(path);
        }
    }

    private static string FindRepoRoot()
    {
        var dir = new DirectoryInfo(AppContext.BaseDirectory);
        while (dir != null)
        {
            if (Directory.Exists(Path.Combine(dir.FullName, "nodejs")))
                return dir.FullName;
            dir = dir.Parent;
        }
        throw new InvalidOperationException("Could not find repository root");
    }

    private static string GetCliPath(string repoRoot)
    {
        var envPath = Environment.GetEnvironmentVariable("COPILOT_CLI_PATH");
        if (!string.IsNullOrEmpty(envPath)) return envPath;

        // As of CLI 1.0.64-1 the @github/copilot package is a thin loader; the
        // runnable index.js ships in the installed platform package
        // (e.g. @github/copilot-linux-x64). Exactly one is installed.
        var githubModules = Path.Join(repoRoot, "nodejs", "node_modules", "@github");
        if (Directory.Exists(githubModules))
        {
            var candidate = Directory.EnumerateDirectories(githubModules, "copilot-*")
                .Select(dir => Path.Join(dir, "index.js"))
                .FirstOrDefault(File.Exists);
            if (candidate != null)
                return candidate;
        }

        throw new InvalidOperationException(
            $"CLI not found under {githubModules}. Run 'npm install' in the nodejs directory first.");
    }

    public async Task ConfigureForTestAsync(string testFile, [CallerMemberName] string? testName = null)
    {
        // Convert test method names to lowercase snake_case for snapshot filenames
        // to avoid case collisions on case-insensitive filesystems (macOS/Windows)
        var sanitizedName = Regex.Replace(testName!, @"[^a-zA-Z0-9]", "_").ToLowerInvariant();
        var snapshotPath = Path.Combine(_repoRoot, "test", "snapshots", testFile, $"{sanitizedName}.yaml");
        await _proxy.ConfigureAsync(snapshotPath, WorkDir);
    }

    public Task<List<ParsedHttpExchange>> GetExchangesAsync()
    {
        return _proxy.GetExchangesAsync();
    }

    public Task SetCopilotUserByTokenAsync(string token, CopilotUserConfig response)
    {
        return _proxy.SetCopilotUserByTokenAsync(token, response);
    }

    public Dictionary<string, string> GetEnvironment()
    {
        var env = Environment.GetEnvironmentVariables()
            .Cast<System.Collections.DictionaryEntry>()
            .ToDictionary(e => (string)e.Key, e => e.Value?.ToString());

        env["COPILOT_API_URL"] = ProxyUrl;
        // Route GitHub API calls (e.g. the MCP registry policy check) to the
        // replay proxy so MCP enablement stays hermetic. Without this the CLI
        // reaches the real api.github.com, which is slow/unreachable on macOS
        // CI runners and makes MCP servers time out before reaching connected.
        env["COPILOT_DEBUG_GITHUB_API_URL"] = ProxyUrl;
        env["COPILOT_HOME"] = HomeDir;
        env["GH_CONFIG_DIR"] = HomeDir;
        env["XDG_CONFIG_HOME"] = HomeDir;
        env["XDG_STATE_HOME"] = HomeDir;
        env["COPILOT_MCP_APPS"] = "true";
        env["MCP_APPS"] = "true";
        if (!string.IsNullOrEmpty(_proxy.ConnectProxyUrl) && !string.IsNullOrEmpty(_proxy.CaFilePath))
        {
            const string noProxy = "127.0.0.1,localhost,::1";
            env["HTTP_PROXY"] = _proxy.ConnectProxyUrl;
            env["HTTPS_PROXY"] = _proxy.ConnectProxyUrl;
            env["http_proxy"] = _proxy.ConnectProxyUrl;
            env["https_proxy"] = _proxy.ConnectProxyUrl;
            env["NO_PROXY"] = noProxy;
            env["no_proxy"] = noProxy;
            env["NODE_EXTRA_CA_CERTS"] = _proxy.CaFilePath;
            env["SSL_CERT_FILE"] = _proxy.CaFilePath;
            env["REQUESTS_CA_BUNDLE"] = _proxy.CaFilePath;
            env["CURL_CA_BUNDLE"] = _proxy.CaFilePath;
            env["GIT_SSL_CAINFO"] = _proxy.CaFilePath;
            env["GH_TOKEN"] = "";
            env["GITHUB_TOKEN"] = "";
            env["GH_ENTERPRISE_TOKEN"] = "";
            env["GITHUB_ENTERPRISE_TOKEN"] = "";
        }

        env["GITHUB_TOKEN"] = env["GH_TOKEN"] = DefaultGitHubToken;

        // Disable HMAC auth for E2E runs. CI sets COPILOT_HMAC_KEY at the job
        // level as an ambient credential, but the replay snapshots are captured
        // against Bearer/OAuth (SDK-token) requests. In stdio the SDK token
        // outranks HMAC so this is a no-op, but in-process auth resolution runs
        // host-side in this process and would otherwise pick HMAC (which ranks
        // above the GitHub token) and fail provider.getEndpoint. An empty value
        // disables the method (runtime filters out empty HMAC keys).
        env["COPILOT_HMAC_KEY"] = "";
        env["CAPI_HMAC_KEY"] = "";

        return env!;
    }

    private static string? GetEffectiveGitHubTokenForTests()
    {
        return Environment.GetEnvironmentVariable("GITHUB_ACTIONS") == "true"
            ? DefaultGitHubToken
            : Environment.GetEnvironmentVariable("GITHUB_TOKEN");
    }

    public CopilotClient CreateClient(
        CopilotClientOptions? options = null,
        bool autoInjectGitHubToken = true,
        bool persistent = false,
        IReadOnlyDictionary<string, string>? environment = null)
    {
        options ??= new CopilotClientOptions();

        options.WorkingDirectory ??= WorkDir;
        options.Logger ??= Logger;

        // Tests must supply environment via the 'environment' parameter, which the
        // harness routes to the right place per transport (the connection for
        // child-process transports, the host process for in-process). Setting
        // options.Environment directly bypasses that routing and is unsupported
        // in-process, so reject it here.
        if (options.Environment is not null)
        {
            throw new ArgumentException(
                "Do not set options.Environment in E2E tests; pass the 'environment' parameter to CreateClient instead.",
                nameof(options));
        }

        // The full environment the client runs with: harness defaults (proxy
        // redirect, isolated home, cleared HMAC/tokens, etc.) unless the test
        // supplied a complete replacement.
        var env = environment is not null
            ? environment.ToDictionary(kvp => kvp.Key, kvp => kvp.Value)
            : GetEnvironment();

        // When the test doesn't pin a transport, leave Connection null so
        // CopilotClient honors COPILOT_SDK_DEFAULT_CONNECTION (stdio by default,
        // or in-process); the CI matrix uses this to run the suite under both.
        // Tests that need a specific transport set options.Connection directly.
        var cliPath = GetCliPath(_repoRoot);
        switch (options.Connection)
        {
            case null when !IsInProcess(null):
                // No explicit connection and not the in-process default: the
                // default resolves to stdio, so materialize it here so the
                // environment can be attached to the connection below.
                options.Connection = RuntimeConnection.ForStdio(path: cliPath);
                break;
            case null:
                // In-process default: leave Connection unset so CopilotClient's
                // ResolveDefaultConnection honors COPILOT_SDK_DEFAULT_CONNECTION.
                break;
            case ChildProcessRuntimeConnection child when child.Path is null:
                child.Path = cliPath;
                break;
        }

        if (IsInProcess(options.Connection))
        {
            // In-process hosting: runtime code runs host-side in this process (the
            // loaded cdylib) and reads the ambient process environment rather than
            // the environment passed to copilot_runtime_host_start, so the per-test
            // redirects, cleared tokens/HMAC, and isolated home must be mirrored
            // onto this process's real environment. Restored after each test by
            // InProcessEnvIsolationAttribute.
            foreach (var (name, value) in env)
            {
                InProcessEnvIsolation.Apply(name, value);
            }
        }
        else if (options.Connection is ChildProcessRuntimeConnection child)
        {
            // Child-process transport: hand the environment to the spawned child
            // via the connection, where per-client environment is coherent.
            child.Environment = env;
        }

        // Auto-inject auth token unless connecting to an existing runtime via URI.
        var isExistingRuntime = options.Connection is UriRuntimeConnection;
        if (autoInjectGitHubToken
            && string.IsNullOrEmpty(options.GitHubToken)
            && !isExistingRuntime)
        {
            options.GitHubToken = GetEffectiveGitHubTokenForTests();
        }

        var client = new CopilotClient(options);
        lock (_clientsLock)
        {
            if (persistent)
            {
                _persistentClients.Add(client);
            }
            else
            {
                _transientClients.Add(client);
            }
        }
        return client;
    }

    public void UntrackClient(CopilotClient client)
    {
        lock (_clientsLock)
        {
            _persistentClients.Remove(client);
            _transientClients.Remove(client);
        }
    }

    public async Task CleanupAfterTestAsync()
    {
        // Per-test cleanup only stops clients created for a specific test.
        // The shared persistent client and temp directories are cleaned when the fixture is disposed.
        var errors = new List<Exception>();
        CopilotClient[] transientClients;

        lock (_clientsLock)
        {
            transientClients = [.. _transientClients];
            _transientClients.Clear();
        }

        foreach (var client in transientClients)
        {
            try
            {
                await StopClientForCleanupAsync(client);
            }
            catch (Exception ex) when (IsTransientCleanupException(ex))
            {
                errors.Add(ex);
            }
        }

        if (errors.Count == 1)
        {
            throw errors[0];
        }
        if (errors.Count > 1)
        {
            throw new AggregateException(errors);
        }
    }

    public async ValueTask DisposeAsync()
    {
        var errors = new List<Exception>();
        CopilotClient[] clients;

        lock (_clientsLock)
        {
            clients = [.. _persistentClients.Concat(_transientClients)];
            _persistentClients.Clear();
            _transientClients.Clear();
        }

        foreach (var client in clients)
        {
            try
            {
                await StopClientForCleanupAsync(client);
            }
            catch (Exception ex) when (IsTransientCleanupException(ex))
            {
                errors.Add(ex);
            }
        }

        // Skip writing snapshots in CI to avoid corrupting them on test failures
        var isCI = !string.IsNullOrEmpty(Environment.GetEnvironmentVariable("GITHUB_ACTIONS"));
        try { await _proxy.StopAsync(skipWritingCache: isCI); } catch (Exception ex) when (IsTransientCleanupException(ex)) { errors.Add(ex); }

        try { await DeleteDirectoryAsync(HomeDir); } catch (Exception ex) when (IsTransientCleanupException(ex)) { errors.Add(ex); }
        try { await DeleteDirectoryAsync(WorkDir); } catch (Exception ex) when (IsTransientCleanupException(ex)) { errors.Add(ex); }

        if (errors.Count == 1)
        {
            throw errors[0];
        }
        if (errors.Count > 1)
        {
            throw new AggregateException(errors);
        }
    }

    private static async Task DeleteDirectoryAsync(string path)
    {
        const int maxAttempts = 40;
        var delay = TimeSpan.FromMilliseconds(50);
        var lastException = (Exception?)null;

        for (var attempt = 1; attempt <= maxAttempts; attempt++)
        {
            if (!Directory.Exists(path))
            {
                return;
            }

            try
            {
                Directory.Delete(path, recursive: true);
                return;
            }
            catch (Exception ex) when (IsTransientCleanupException(ex))
            {
                lastException = ex;
                if (attempt == maxAttempts)
                {
                    break;
                }

                await Task.Delay(delay);
                delay = TimeSpan.FromMilliseconds(Math.Min(delay.TotalMilliseconds * 2, 250));
            }
        }

        if (Directory.Exists(path))
        {
            throw new IOException($"Failed to delete directory '{path}' after {maxAttempts} attempts.", lastException);
        }
    }

    /// <summary>
    /// Determines whether the resolved transport is the in-process (FFI) host,
    /// mirroring <see cref="CopilotClient"/>'s own default-connection resolution:
    /// an explicit <see cref="InProcessRuntimeConnection"/>, or (when no connection
    /// is given) the COPILOT_SDK_DEFAULT_CONNECTION=inprocess default.
    /// </summary>
    private static bool IsInProcess(RuntimeConnection? connection)
    {
        if (connection is InProcessRuntimeConnection)
        {
            return true;
        }
        if (connection is null)
        {
            return string.Equals(
                Environment.GetEnvironmentVariable("COPILOT_SDK_DEFAULT_CONNECTION"),
                "inprocess",
                StringComparison.OrdinalIgnoreCase);
        }
        return false;
    }

    // Inproc holds the session-store SQLite handle in-process; graceful StopAsync releases it so the temp-dir delete succeeds on Windows.
    private static async Task StopClientForCleanupAsync(CopilotClient client)
    {
        var isInProcess = string.Equals(
            Environment.GetEnvironmentVariable("COPILOT_SDK_DEFAULT_CONNECTION"),
            "inprocess",
            StringComparison.OrdinalIgnoreCase);
        if (isInProcess)
        {
            await client.StopAsync();
        }
        else
        {
            await client.ForceStopAsync();
        }
    }

    private static bool IsTransientCleanupException(Exception exception)
        => exception is IOException or UnauthorizedAccessException;
}
